frontend/src/metabase/nav/components/TrialBanner/TrialBanner.tsx

```diff
@@ -0,0 +1,79 @@
+import dayjs from "dayjs";
+import { useMemo } from "react";
+import { msgid, ngettext, t } from "ttag";
+
+import { useUserSetting } from "metabase/common/hooks";
+import { Banner } from "metabase/components/Banner";
+import ExternalLink from "metabase/core/components/ExternalLink";
+import CS from "metabase/css/core/index.css";
+import { getStoreUrl } from "metabase/selectors/settings";
+import { Flex, Text } from "metabase/ui";
+import type { TokenStatus } from "metabase-types/api";
+export const TrialBanner = ({ tokenStatus }: { tokenStatus: TokenStatus }) => {
+  const [lastDismissed, setLastDismissed] = useUserSetting(
+    "trial-banner-dismissal-timestamp",
+  );
+
+  const now = dayjs();
+  const tokenExpiryDate = dayjs(tokenStatus["valid-thru"]);
+
+  const daysRemaining = tokenExpiryDate.diff(now, "day");
+  const lastDay = daysRemaining === 0;
+
+  const shouldShowBanner = useMemo(() => {
+    // No banner if the trial already expired
+    if (daysRemaining < 0) {
+      return false;
+    }
+
+    // In the last 3 days, check dismissal logic with stricter reappearance
+    if (daysRemaining <= 3) {
+      const bannerReappearanceThreshold = tokenExpiryDate
+        .subtract(daysRemaining, "day")
+        .unix();
+
+      // Banner shows if it hasn't been dismissed after the reappearance point
+      return !lastDismissed || lastDismissed < bannerReappearanceThreshold;
+    }
+
+    // For the first 11 days, check if it was ever dismissed
+    return !lastDismissed;
```

Comment:
The contents of this `useMemo` should be a separate function in a `utils` file with the signature:

```ts
function shouldShowBanner(tokenExpiryDate, lastDismissed)
```

That way you can test nearly all the complex logic in this component with very simple unit tests without the need for mocking settings and whatnot.